src/items.rs

```diff
@@ -1164,25 +1215,40 @@ pub(crate) fn format_trait(
         rewrite_generics(context, rewrite_ident(context, item.ident), generics, shape)?;
     result.push_str(&generics_str);
 
-    // FIXME(#2055): rustfmt fails to format when there are comments between trait bounds.
     if !bounds.is_empty() {
-        // Retrieve *unnormalized* ident (See #6069)
-        let source_ident = context.snippet(item.ident.span);
-        let ident_hi = context.snippet_provider.span_after(item.span, source_ident);
-        let bound_hi = bounds.last().unwrap().span().hi();
-        let snippet = context.snippet(mk_sp(ident_hi, bound_hi));
-        if contains_comment(snippet) {
-            return None;
-        }
+        if context.config.version() == Version::Two {
+            let after_colon = context
+                .snippet_provider
+                .span_after(item.span.with_lo(item.ident.span.hi()), ":");
 
-        result = rewrite_assign_rhs_with(
-            context,
-            result + ":",
-            bounds,
-            shape,
-            &RhsAssignKind::Bounds,
-            RhsTactics::ForceNextLineWithoutIndent,
-        )?;
+            let span = mk_sp(after_colon, body_lo);
+            let shape = if result.contains('\n') {
+                shape
+            } else {
+                // `offset_left` takes into account what we've rewritten already + 1 for `:`
+                // `sub_width` take into account the trailing `{`
+                shape.offset_left(header.width() + 1)?.sub_width(1)?
```

Comment:
Might need to be `header.width() + 2` if we're applying [`space_before_colon`](https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#space_before_colon), Also I'm realizing that the header doesn't contain any rewritten generics. Might be better to move the shape derivation into `rewrite_bounds`